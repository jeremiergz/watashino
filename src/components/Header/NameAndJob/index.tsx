import Heading from 'components/Heading';
import Image from 'components/Image';
import Link from 'components/Link';
import FlexBox from 'components/primitives/FlexBox';
import { graphql, useStaticQuery } from 'gatsby';
import React from 'react';
import { Routes } from 'utils/Routes';

const NameAndJob: React.FC = () => {
  const { dataJson, file } = useStaticQuery(graphql`
    query HeaderData {
      dataJson {
        jobTitle
      }
      file(relativePath: { eq: "profile-pic.png" }) {
        childImageSharp {
          fixed(height: 80, quality: 100, width: 80) {
            ...GatsbyImageSharpFixed_withWebp
          }
        }
      }
    }
  `);
  return (
    <Link
      alignItems="center"
      display="flex"
      flexDirection={{ _: 'column', tablet: 'row' }}
      justifyContent={{ _: 'center', tablet: 'flex-start' }}
      textAlign={{ _: 'center', tablet: 'left' }}
      to={Routes.aboutMe}
    >
      <Image fixed={file.childImageSharp.fixed} imgStyle={{ height: 80, width: 80 }} />
      <FlexBox flexDirection="column" justifyContent="center" marginLeft={{ tablet: 2 }}>
        <Heading color="primary" fontWeight="bold" letterSpacing={-1} margin={0} padding={2} variant="h3" zIndex={1}>
          Jeremie Rodriguez
        </Heading>
        <Heading
          fontWeight="semi-bold"
          margin={0}
          marginTop={-2}
          paddingBottom={2}
          paddingTop={0}
          paddingX={2}
          variant="h4"
        >
          {dataJson.jobTitle}
        </Heading>
      </FlexBox>
    </Link>
  );
};

NameAndJob.displayName = 'NameAndJob';

export default NameAndJob;
