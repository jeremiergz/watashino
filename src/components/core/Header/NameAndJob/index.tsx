import { graphql, useStaticQuery } from 'gatsby';
import Img from 'gatsby-image';
import React from 'react';
import { Routes } from '../../../../utils/Routes';
import Heading from '../../../common/Heading';
import Link from '../../../common/Link';
import Flex from '../../../primitives/Flex';

const NameAndJob = () => {
  const { dataJson, file } = useStaticQuery<GraphQL.HeaderDataQuery>(graphql`
    query HeaderData {
      dataJson {
        jobTitle
      }
      file(relativePath: { eq: "profile-pic.png" }) {
        childImageSharp {
          fixed(height: 80, width: 80) {
            ...GatsbyImageSharpFixed_withWebp
          }
        }
      }
    }
  `);
  return (
    <Link
      alignItems="center"
      display="flex"
      flexDirection={{ _: 'column', tablet: 'row' }}
      justifyContent={{ _: 'center', tablet: 'flex-start' }}
      textAlign={{ _: 'center', tablet: 'left' }}
      to={Routes.aboutMe}
    >
      <Img fixed={file.childImageSharp.fixed} imgStyle={{ height: 80, width: 80 }} />
      <Flex flexDirection="column" justifyContent="center" marginLeft={{ tablet: 2 }}>
        <Heading color="primary" fontWeight="bold" letterSpacing={-1} margin={0} padding={2} variant="h3" zIndex={1}>
          Jeremie Rodriguez
        </Heading>
        <Heading
          fontWeight="semi-bold"
          margin={0}
          marginTop={-2}
          paddingBottom={2}
          paddingTop={0}
          paddingX={2}
          variant="h4"
        >
          {dataJson.jobTitle}
        </Heading>
      </Flex>
    </Link>
  );
};

NameAndJob.displayName = 'NameAndJob';

export default NameAndJob;
