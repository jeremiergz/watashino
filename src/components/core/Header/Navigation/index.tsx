import { graphql, useStaticQuery } from 'gatsby';
import React, { ReactNode } from 'react';
import Anchor from '../../../common/Anchor';
import Link from '../../../common/Link';
import Box from '../../../primitives/Box';
import Flex from '../../../primitives/Flex';
import BookIcon from '../../../svg/Book';
import DownloadIcon from '../../../svg/Download';
import HouseIcon from '../../../svg/House';
import MoodIcon from '../../../svg/Mood';
import styled from 'styled-components';

const iconsMapping = {
  book: BookIcon,
  home: HouseIcon,
  mood: MoodIcon,
};

const LinkContainer = ({ children, ...rest }: { children: ReactNode }) => (
  <Flex alignItems="center" flexDirection="column" justifyContent="center" minWidth={96} {...rest}>
    {children}
  </Flex>
);

const Svg = styled(Box)`
  fill: ${({ theme }) => theme.svg};
`;

const Navigation = () => {
  const {
    allMarkdownRemark: { totalCount },
    allNavigationJson: { nodes: links },
    dataJson: { resumeFile },
  } = useStaticQuery<GraphQL.NavigationDataQuery>(graphql`
    query NavigationData {
      allMarkdownRemark {
        totalCount
      }
      allNavigationJson {
        nodes {
          icon
          ignoreInNavigation
          keywords
          name
          navOrder
          to
        }
      }
      dataJson {
        resumeFile
      }
    }
  `);
  return (
    <Flex as="nav" marginTop={{ _: 16, tablet: 0 }}>
      {links
        .filter(link => {
          if (link.to === '/posts' && totalCount === 0) return false;
          return !link.ignoreInNavigation;
        })
        .sort((a, b) => a.navOrder - b.navOrder)
        .map(link => (
          <LinkContainer key={link.name}>
            <Link alignItems="center" display="flex" flexDirection="column" partiallyActive to={link.to}>
              <Svg as={iconsMapping[link.icon]} height={32} width={32} />
              {link.name}
            </Link>
          </LinkContainer>
        ))}
      <LinkContainer>
        <Anchor alignItems="center" display="flex" download flexDirection="column" href={resumeFile}>
          <Svg as={DownloadIcon} height={32} width={32} />
          Resume
        </Anchor>
      </LinkContainer>
    </Flex>
  );
};

Navigation.displayName = 'Navigation';

export default Navigation;
