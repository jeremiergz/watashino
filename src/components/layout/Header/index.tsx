import Conditional from '@/components/common/Conditional';
import clsx from 'clsx';
import React from 'react';

type Content = 'whereami' | 'whoami';

const navItems: { label: Content }[] = [{ label: 'whoami' }, { label: 'whereami' }];

function Header({ current, onCurrentChange }: HeaderProps): JSX.Element {
  const handleCurrentChange = (newCurrent: Content) => () => {
    if (typeof onCurrentChange === 'function' && newCurrent !== current) {
      onCurrentChange(newCurrent);
    }
  };

  return (
    <header className="fixed top-0 left-0 right-0 z-50 flex h-12 w-full max-w-5xl justify-center bg-gray-900 md:relative md:h-16 md:justify-end md:bg-transparent">
      <nav className="flex font-semibold">
        {navItems.map((item, index) => {
          const labelStart = item.label.substring(0, item.label.length - 3);
          const labelEnd = item.label.substring(item.label.length - 3, item.label.length);

          return (
            <div className={clsx('relative flex items-center', index !== 0 && 'ml-10')} key={item.label}>
              <Conditional
                condition={item.label === current}
                wrapper={children => (
                  <>
                    <span className="absolute text-sm font-bold text-cyan-600">&gt;</span>
                    {children}
                  </>
                )}
              >
                <button
                  aria-label={`Go to ${item.label}`}
                  className="cursor-pointer rounded py-1 px-4 transition hover:shadow-md hover:brightness-90"
                  onClick={handleCurrentChange(item.label)}
                >
                  <span>{labelStart}</span>
                  <span className="text-cyan-600">{labelEnd}</span>
                </button>
              </Conditional>
            </div>
          );
        })}
      </nav>
    </header>
  );
}

export type { Content };
export type HeaderProps = {
  current: Content;
  onCurrentChange?: (value: Content) => void;
};
export default Header;
