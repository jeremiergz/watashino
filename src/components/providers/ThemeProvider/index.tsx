import React, { createContext, PropsWithChildren, useContext, useEffect, useState } from 'react';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';
import { DarkTheme, LightTheme, Theme } from 'theme';

const defaultMode = 'light';
const localStorageKey = 'theme:mode';

const ThemeContext = createContext<{ mode: ThemeMode; theme: Theme; toggle: () => void }>({
  mode: defaultMode,
  theme: LightTheme,
  toggle: () => null,
});

const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
  const isBrowser = typeof window !== 'undefined';
  const [isInitialized, setIsInitialized] = useState(false);
  const [mode, setMode] = useState<ThemeMode>(defaultMode);
  const toggle = () => {
    const newMode = mode === 'light' ? 'dark' : 'light';
    setMode(newMode);
  };
  const theme = mode === 'light' ? LightTheme : DarkTheme;
  useEffect(() => {
    if (isBrowser && isInitialized) localStorage.setItem(localStorageKey, mode);
  }, [mode]);
  useEffect(() => {
    const storedMode = (isBrowser && (localStorage.getItem(localStorageKey) as ThemeMode)) || defaultMode;
    setMode(storedMode);
    setIsInitialized(true);
  }, []);
  return (
    <ThemeContext.Provider value={{ mode, theme, toggle }}>
      <StyledThemeProvider theme={theme}>{children}</StyledThemeProvider>
    </ThemeContext.Provider>
  );
};

ThemeProvider.displayName = 'ThemeProvider';

/**
 * Returns the theming context that allows retrieval of theme & toggling theme mode.
 */
function useTheming() {
  return useContext(ThemeContext);
}

export { useTheming };
export type ThemeProviderProps = PropsWithChildren<{}>;
export default ThemeProvider;
