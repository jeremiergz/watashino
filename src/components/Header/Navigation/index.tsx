import Link from 'components/Link';
import FlexBox from 'components/primitives/FlexBox';
import { useTheme } from 'components/providers/ThemeProvider';
import SVG from 'components/SVG';
import BookIcon from 'components/svgs/icons/Book';
import DescriptionIcon from 'components/svgs/icons/Description';
import HouseIcon from 'components/svgs/icons/House';
import MoodIcon from 'components/svgs/icons/Mood';
import { graphql, useStaticQuery } from 'gatsby';
import { trackCustomEvent } from 'gatsby-plugin-google-analytics';
import React from 'react';
import { Routes } from 'utils/Routes';

const iconsMapping = {
  book: BookIcon,
  home: HouseIcon,
  mood: MoodIcon,
};

const Navigation: React.FC = () => {
  const { theme } = useTheme();
  const {
    allMarkdownRemark: { totalCount },
    allNavigationJson: { nodes: links },
    dataJson: { resumeLink },
  } = useStaticQuery<GraphQL.NavigationDataQuery>(graphql`
    query NavigationData {
      allMarkdownRemark {
        totalCount
      }
      allNavigationJson {
        nodes {
          icon
          ignoreInNavigation
          keywords
          name
          navOrder
          to
        }
      }
      dataJson {
        resumeLink
      }
    }
  `);
  const handleResumeClick = () => {
    trackCustomEvent({
      action: 'click',
      category: 'Open Resume',
    });
  };
  return (
    <FlexBox as="nav" flexWrap="wrap" justifyContent="center" marginTop={{ _: 16, tablet: 0 }}>
      {links
        .filter(link => {
          if (link.to === Routes.posts && totalCount === 0) return false;
          return !link.ignoreInNavigation;
        })
        .sort((a, b) => a.navOrder - b.navOrder)
        .map(link => (
          <FlexBox alignItems="center" flexDirection="column" justifyContent="center" key={link.name} minWidth={96}>
            <Link alignItems="center" display="flex" flexDirection="column" partiallyActive to={link.to}>
              <SVG as={iconsMapping[link.icon]} fill={theme.colors.text} />
              {link.name}
            </Link>
          </FlexBox>
        ))}
      <FlexBox alignItems="center" flexDirection="column" justifyContent="center" minWidth={96}>
        <Link
          alignItems="center"
          display="flex"
          external
          flexDirection="column"
          onClick={handleResumeClick}
          to={resumeLink}
        >
          <SVG as={DescriptionIcon} fill={theme.colors.text} />
          Resume
        </Link>
      </FlexBox>
    </FlexBox>
  );
};

Navigation.displayName = 'Navigation';

export default Navigation;
