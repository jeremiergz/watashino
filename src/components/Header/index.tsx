import FlexBox from 'components/primitives/FlexBox';
import { useTheme } from 'components/providers/ThemeProvider';
import React from 'react';
import NameAndJob from './NameAndJob';
import Navigation from './Navigation';

const Header: React.FC = () => {
  const { theme } = useTheme();
  return (
    <FlexBox
      alignItems="center"
      as="header"
      flexDirection={{ _: 'column', tablet: 'row' }}
      justifyContent={{ _: 'flex-start', tablet: 'space-between' }}
      margin="auto"
      maxWidth={theme.breakpoints[5]}
      paddingBottom={32}
      paddingTop={{ _: 32, tablet: 48 }}
      paddingX={{ _: 16, tablet: 32 }}
    >
      <NameAndJob />
      <Navigation />
    </FlexBox>
  );
};

Header.displayName = 'Header';

export default Header;
