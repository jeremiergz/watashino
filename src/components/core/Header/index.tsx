import React from 'react';
import useTheme from '../../../hooks/useTheme';
import Flex from '../../primitives/Flex';
import NameAndJob from './NameAndJob';
import Navigation from './Navigation';

const Header = () => {
  const theme = useTheme();
  return (
    <Flex
      alignItems="center"
      as="header"
      flexDirection={{ _: 'column', tablet: 'row' }}
      justifyContent={{ _: 'flex-start', tablet: 'space-between' }}
      margin="auto"
      maxWidth={theme.breakpoints[5]}
      paddingX={{ _: 16, tablet: 32 }}
      paddingY={32}
    >
      <NameAndJob />
      <Navigation />
    </Flex>
  );
};

Header.displayName = 'Header';

export default Header;
