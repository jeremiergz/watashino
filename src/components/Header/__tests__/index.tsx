import ThemeProvider from 'components/providers/ThemeProvider';
import * as Gatsby from 'gatsby';
import React from 'react';
import ReactTestRenderer from 'react-test-renderer';
import Header from '..';

describe('components/Header Test Suite', () => {
  beforeAll(() => {
    const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');
    useStaticQuery.mockImplementation(() => ({
      allMarkdownRemark: {
        totalCount: 1,
      },
      allNavigationJson: {
        nodes: [
          {
            icon: 'icon',
            ignoreInNavigation: false,
            keywords: [],
            name: 'Test',
            navOrder: 0,
            to: '/',
          },
        ],
      },
      dataJson: {
        jobTitle: 'Full Stack Engineer',
      },
      file: {
        childImageSharp: {
          fixed: {
            base64: 'data:image/png;base64',
            aspectRatio: 1,
            src: '/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png',
            srcSet:
              '/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png 1x,\n/static/89ee547c5db6cbf9b37a682e8b3f3f57/1e576/profile-pic.png 1.5x,\n/static/89ee547c5db6cbf9b37a682e8b3f3f57/6cb10/profile-pic.png 2x',
          },
        },
      },
      themeJson: {
        colors: {
          black: '#212121',
          dark: '#424242',
          error: '#f44336',
          gray: '#757575',
          info: '#2196f3',
          lightgray: '#e0e0e0',
          primary: '#31859a',
          secondary: '#17365c',
          success: '#4caf50',
          vsCodeDark: '#1e1e1e',
          white: '#fafafa',
        },
        fonts: {
          main: "'Open Sans', sans-serif",
        },
        fontWeights: {
          bold: 800,
          regular: 400,
          semiBold: 600,
        },
      },
    }));
  });

  it('matches snapshot', () => {
    const jsx = (
      <ThemeProvider>
        <Header />
      </ThemeProvider>
    );
    expect(ReactTestRenderer.create(jsx).toJSON()).toMatchSnapshot();
  });
});
