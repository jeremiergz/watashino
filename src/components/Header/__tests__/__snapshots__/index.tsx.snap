// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/core/Header Test Suite matches snapshot 1`] = `
<header
  className="sc-AxjAm ceTnwe"
  display="flex"
>
  <a
    className="sc-AxjAm sc-AxhUy ghFdRF"
    href="/about-me"
    target="_blank"
  >
    <div
      className="sc-AxhCb eYhTuB gatsby-image-wrapper"
      style={
        Object {
          "display": "inline-block",
          "height": undefined,
          "overflow": "hidden",
          "position": "relative",
          "width": undefined,
        }
      }
    >
      <img
        alt=""
        aria-hidden={true}
        src="data:image/png;base64"
        style={
          Object {
            "height": 80,
            "left": 0,
            "objectFit": "cover",
            "objectPosition": "center",
            "opacity": 1,
            "position": "absolute",
            "top": 0,
            "transitionDelay": "500ms",
            "width": 80,
          }
        }
      />
      <picture>
        <source
          srcSet="/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png 1x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/1e576/profile-pic.png 1.5x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/6cb10/profile-pic.png 2x"
        />
        <img
          alt=""
          loading="lazy"
          onLoad={[Function]}
          src="/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png"
          srcSet="/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png 1x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/1e576/profile-pic.png 1.5x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/6cb10/profile-pic.png 2x"
          style={
            Object {
              "height": 80,
              "left": 0,
              "objectFit": "cover",
              "objectPosition": "center",
              "opacity": 0,
              "position": "absolute",
              "top": 0,
              "transition": "opacity 500ms",
              "width": 80,
            }
          }
        />
      </picture>
      <noscript
        dangerouslySetInnerHTML={
          Object {
            "__html": "<picture><source srcset=\\"/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png 1x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/1e576/profile-pic.png 1.5x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/6cb10/profile-pic.png 2x\\" /><img loading=\\"lazy\\" srcset=\\"/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png 1x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/1e576/profile-pic.png 1.5x,
/static/89ee547c5db6cbf9b37a682e8b3f3f57/6cb10/profile-pic.png 2x\\" src=\\"/static/89ee547c5db6cbf9b37a682e8b3f3f57/8b091/profile-pic.png\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
          }
        }
      />
    </div>
    <div
      className="sc-AxjAm inuzPJ"
      display="flex"
    >
      <span
        className="sc-AxirZ sc-AxiKw ebUoHp"
        color="primary"
        fontWeight="bold"
        letterSpacing={-1}
      >
        Jeremie Rodriguez
      </span>
      <span
        className="sc-AxirZ sc-AxiKw bHICOK"
        fontWeight="semi-bold"
      >
        Full Stack Engineer
      </span>
    </div>
  </a>
  <nav
    className="sc-AxjAm fbAfJg"
    display="flex"
  >
    <div
      className="sc-AxjAm gqPulF"
      display="flex"
    >
      <a
        className="sc-AxjAm sc-AxhUy iRGypK"
        href="/"
        target="_blank"
      >
        <svg
          className="sc-AxgMl fvMXKT"
          fill="#262626"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        />
        Test
      </a>
    </div>
    <div
      className="sc-AxjAm gqPulF"
      display="flex"
    >
      <a
        className="sc-AxjAm sc-AxhUy iRGypK"
        onClick={[Function]}
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          className="sc-AxgMl fvMXKT sc-AxgMl fvMXKT"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            fill="#262626"
          />
        </svg>
        Resume
      </a>
    </div>
  </nav>
</header>
`;
